[#20]feat: 회원 탈퇴 기능

Yo0oN opened the pull request:
회원 탈퇴 기능 구현.
날짜 확인 배치 클래스를 구현하여 회원 탈퇴 1년 후 관련 정보를 삭제하는 기능 추가 예정

[#20] 회원 탈퇴 기능

Yo0oN committed:
[#20]feat: 회원 탈퇴 기능

회원 탈퇴 기능 구현.
날짜 확인 배치 클래스를 이용하여 회원 탈퇴 1년 후 관련 정보를 삭제할 예정

[#20] 회원 탈퇴 기능

Jane096 commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -114,4 +114,21 @@
+     * 회원 탈퇴 기능
+     * @param memberDTO
+     * @return
+     */
+    @PostMapping(value = "/memberWithdraw")
+    public ResponseEntity<HttpStatus> memberWithdraw(@RequestBody @Valid MemberDTO memberDTO) {
```
이 부분은 그냥 `/withdraw` 로 끝내도 `POST /members/withdraw` 로 의미가 통할 것 같습니다!

f-lab-dev replied:
적극적인 리뷰 좋습니다~ 다만 `/members/{userId}` URL과 중복되지 않을까요? 이런 경우엔 어떻게 설계하는게 좋을까요?

f-lab-dev replied:
그리고 탈퇴를 하는것은 결국 회원의 리소스를 삭제하는 것인데 `POST`를 사용하는게 맞을까요?
이거를 한번 읽어주세요~

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

Yo0oN replied:
회원 리소스를 삭제하는것이니 `DeleteMapping`으로 수정하겠습니다!
URL은 `/withdraw/{userId}' 로 하는건 어떨까요?

Jane096 commented on src/main/java/com/festa/service/MemberService.java:
```diff
@@ -18,4 +18,6 @@ public void insertMemberInfo(MemberDTO memberDTO) {
     public boolean isUserIdExist(long userId) {
         return memberDAO.isUserIdExist(userId);
     }
+
+    public void memberWithdraw(MemberDTO memberDTO) { memberDAO.modifyMemberInfoForWithdraw(memberDTO); }
```
이 부분은 인텔리제이에서 자동으로 이렇게 만들어진 것 같은데 한줄 보다는 개행이 나을 것 같습니다~!

f-lab-dev commented on src/main/resources/mapper/MemberMapper.xml:
```diff
@@ -35,4 +35,11 @@
+        update members
+        set memberWithdraw = TRUE
+        and memberWithdrawDate = NOW()
+        where userId = #{userId};
```
`where`, `update` 같은 예약어들은 대문자로 표현해주시면 가독성을 조금 더 올릴 수 있습니다~

Yo0oN replied:
네 수정하겠습니다~

Yo0oN committed:
[#20]feat:회원 탈퇴 기능 수정

삭제 기능이므로 PostMapping 대신 DeleteMapping으로 변경
MemberMapper.xml 예약어는 대문자로 변경

[#20] 회원 탈퇴 기능

f-lab-dev commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -120,7 +115,7 @@
      * @return
      */
-    @PostMapping(value = "/memberWithdraw")
+    @DeleteMapping(value = "/{userId}/withdraw")
```
1. `DELETE /{userId}` 만으로도 특정 유저를 삭제한다라는 행위가 충분히 표현되지 않을까요?
2. 유저 탈퇴는 보통 자기 스스로 탈퇴하는겁니다. 그러면 굳이 `path variable`로 `userId`를 입력받을 필요가 있을까요??

Yo0oN replied:
이후에 참가 신청한 이벤트를 취소하거나, 등록 이벤트를 제거할 때에도 DELETE가 사용될거같습니다. 그때와 구분을 위해서 withdraw는 남겨놔야 할은데 DELETE /withdraw 로 해도 괜찮을까요?

f-lab-dev replied:
이벤트를 삭제할때는 앞의 주소가 달라지지 않을까요? `DELETE /my-events/{eventId}`

Yo0oN replied:
그렇게 된다면 회원 삭제는 DELETE / 로만 해도 괜찮을것같습니다! 수정하겠습니다

Yo0oN committed:
[#20]feat:회원 탈퇴 기능

회원 탈퇴 시 DELETE /{userid}/withdraw 대신 DELETE / 로 수정.
{userid}로 입력받지 않더라도 MemberDTO로 멤버 정보를 알 수 있고, DELETE의 경우 이벤트 신청 취소, 이벤트 제거 정도 시에만 사용하기 때문에 DELETE / 로만 사용하여도 어떤 행위를 하는지 표현할 수 있다.

[#20] 회원 탈퇴 기능
